lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,476 @@
+import signal
+import subprocess
+import sys
+import textwrap
+from pathlib import Path
+from typing import List
+
+from dbus_fast.aio import MessageBus
+from dbus_fast.constants import BusType, MessageType
+from dbus_fast.errors import DBusError
+from dbus_fast.message import Message
+from ops.charm import CharmBase
+from ops.framework import EventBase
+
+LIBID = "HPCTEAMUSEONLY"
+LIBAPI = 0
+LIBPATCH = 1
+PYDEPS = ["dbus-fast>=1.90.2"]
+
+_logger = logging.getLogger(__name__)
+_juju_unit = None
+_service_states = {}
+_service_hook_regex_filter = re.compile(r"service-(?P<service>[\w\\:-]*)-(?:started|stopped)")
+_DBUS_CHAR_MAPPINGS = {
+    "_5f": "_",  # _ must be first since char mappings contain _.
+    "_40": "@",
+    "_2e": ".",
+    "_2d": "-",
+    "_5c": "\\",
+}
+
+
+def _systemctl(*args) -> None:
+    """Control systemd by via executed `systemctl ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if systemctl command fails.
+    """
+    cmd = ["systemctl", *args]
+    _logger.debug(f"systemd: Executing command {cmd}")
+    try:
+        subprocess.check_output(cmd)
+    except subprocess.CalledProcessError as e:
+        _logger.error(f"systemctl command failed: {e}")
+        raise
+
+
+_daemon_reload = functools.partial(_systemctl, "daemon-reload")
+_start_service = functools.partial(_systemctl, "start")
+_stop_service = functools.partial(_systemctl, "stop")
+_enable_service = functools.partial(_systemctl, "enable")
+_disable_service = functools.partial(_systemctl, "disable")
+
+
+class ServiceStartedEvent(EventBase):
+    """Event emitted when service has started."""
+
+
+class ServiceStoppedEvent(EventBase):
+    """Event emitted when service has stopped."""
+
+
+class SystemdNotices:
+    """Observe systemd services on your machine base."""
+
+    def __init__(self, charm: CharmBase, services: List[str]) -> None:
+        """Instantiate systemd notices service."""
+        self._charm = charm
+        self._services = services
+        unit_name = self._charm.unit.name.replace("/", "-")
+        self._service_file = Path(f"/etc/systemd/system/juju-{unit_name}-systemd-notices.service")
+
+        _logger.debug(f"Attaching systemd notice events to charm {self._charm.__class__.__name__}")
+        for service in self._services:
+            self._charm.on.define_event(f"service_{service}_started", ServiceStartedEvent)
+            self._charm.on.define_event(f"service_{service}_stopped", ServiceStoppedEvent)
+
+    def subscribe(self) -> None:
+        """Subscribe charmed operator to observe status of systemd services."""
+        _logger.debug(f"Generating systemd notice hooks for {self._services}")
+        for hook in [
+            Path(f"hooks/service-{service}-{hook_type}")
+            for service in self._services
+            for hook_type in ("started", "stopped")
+        ]:
+            if hook.exists():
+                _logger.debug(f"Hook {hook.name} already exists. Skipping...")
```

Comment:
```suggestion
                _logger.debug("Hook %s already exists. Skipping...", hook.name)
```

Reply:
Refactored in commit https://github.com/canonical/operator-libs-linux/pull/108/commits/3bd603a65b70ec45dfbc4ceff82114739179f9c6